ejb3/src/main/java/org/jboss/as/ejb3/component/stateful/StatefulSessionSynchronizationInterceptor.java

```diff
@@ -139,6 +189,13 @@ public Object processInvocation(final InterceptorContext context) throws Excepti
                         instance.getComponent().getCache().release(instance);
                     }
                 }
+
+                if (instance.delayedAfterCompletion().compareAndSet(true, false)) {
+                    final int status = instance.afterCompletionStatus().get();
+                    executeAfterCompletion(instance, status);
+                } else {
+                    instance.delayedAfterCompletion().set(false);
```

Comment:
I think this cannot do anything useful.  If the CAS fails, that means the variable was already set to ``false``; if someone set it to ``true`` in the meantime, you will just overwrite that value blindly.